Add guessing support

sekikn opened the pull request:
Hi, I'm thinking guessing file type and column definitions from Avro's file header would be useful. Is it possible to merge this PR by any chance?
It seems to work in my environment, as follows:

Define minimal configuration:

```
$ cat /tmp/config.yml 
in:
  type: file
  path_prefix: /home/sekikn/repos/embulk-parser-avro/src/test/resources/org/embulk/parser/avro/items.avro
```

Build this plugin and install the gem:

```
$ ./gradlew gem
$ embulk gem install pkg/embulk-parser-avro-0.3.0.gem 
2021-01-12 11:30:43.844 +0900: Embulk v0.9.23

Gem plugin path is: /home/sekikn/.embulk/lib/gems

Successfully installed embulk-parser-avro-0.3.0
1 gem installed
```

Make sure guessing works:

```
$ embulk guess /tmp/config.yml -g avro
2021-01-12 11:31:35.543 +0900: Embulk v0.9.23
2021-01-12 11:31:36.034 +0900 [WARN] (main): DEPRECATION: JRuby org.jruby.embed.ScriptingContainer is directly injected.
2021-01-12 11:31:37.466 +0900 [INFO] (main): Gem's home and path are set by default: "/home/sekikn/.embulk/lib/gems"
2021-01-12 11:31:37.984 +0900 [INFO] (main): Started Embulk v0.9.23
2021-01-12 11:31:38.064 +0900 [INFO] (0001:guess): Listing local files at directory '/home/sekikn/repos/embulk-parser-avro/src/test/resources/org/embulk/parser/avro' filtering filename by prefix 'items.avro'
2021-01-12 11:31:38.065 +0900 [INFO] (0001:guess): "follow_symlinks" is set false. Note that symbolic links to directories are skipped.
2021-01-12 11:31:38.066 +0900 [INFO] (0001:guess): Loading files [/home/sekikn/repos/embulk-parser-avro/src/test/resources/org/embulk/parser/avro/items.avro]
2021-01-12 11:31:38.075 +0900 [INFO] (0001:guess): Try to read 32,768 bytes from input source
2021-01-12 11:31:38.123 +0900 [INFO] (0001:guess): Loaded plugin embulk (0.9.23)
2021-01-12 11:31:38.138 +0900 [INFO] (0001:guess): Loaded plugin embulk (0.9.23)
2021-01-12 11:31:38.155 +0900 [INFO] (0001:guess): Loaded plugin embulk (0.9.23)
2021-01-12 11:31:38.165 +0900 [INFO] (0001:guess): Loaded plugin embulk (0.9.23)
2021-01-12 11:31:38.183 +0900 [INFO] (0001:guess): Loaded plugin embulk-parser-avro (0.3.0)
in:
  type: file
  path_prefix: /home/sekikn/repos/embulk-parser-avro/src/test/resources/org/embulk/parser/avro/items.avro
  parser:
    charset: UTF-8
    newline: LF
    type: avro
    columns:
    - {name: id, type: long}
    - {name: code, type: long}
    - {name: name, type: string}
    - {name: description, type: string}
    - {name: flag, type: boolean}
    - {name: created_at, type: string}
    - {name: created_at_utc, type: double}
    - {name: price, type: double}
    - {name: spec, type: json}
    - {name: tags, type: json}
    - {name: options, type: json}
    - {name: item_type, type: string}
    - {name: dummy, type: string}
```

sekikn committed:
Add guessing support

joker1007 commented on lib/embulk/guess/avro.rb:
```diff
@@ -1,61 +1,57 @@
-    #    end
-    #  end
-    #end
+        def convert_type(field)
```
In general, Ruby programmers don't nest method definitions except for special needs.
In this case, I think that this method can be defined in a standard position.

joker1007 commented on lib/embulk/guess/avro.rb:
```diff
@@ -1,61 +1,57 @@
-    #  end
-    #end
+        def convert_type(field)
+          type_map = {
```
This value is invariant and it doesn't depend on other variables.
I think that it should be defined as a constant of this class.

joker1007 reviewed (commented):
Thank you for your contribution!!
This feature is very nice.

I have some comments about not Ruby-ish codes.
Would you fix them?

joker1007 commented:
And if possible, I would like it to be written in Java.
It is because the base of this plugin is implemented in Java, and JRuby is likely to be optional in Embulk in the future.

If it is difficult, I will include it as Ruby, but I may rewrite it in Java in the future.

sekikn committed:
Rewrite the guess plugin with Java

sekikn commented:
Thank you for the review @joker1007!

> And if possible, I would like it to be written in Java.

To be honest, I usually don't write Ruby code so much and am more familar with Java, so I'd welcome your suggestion ;)
Please wait for updating the PR a few days.

sekikn committed:
Merge remote-tracking branch 'upstream/master' into guessing-support

sekikn commented:
Updated the PR. In addition, #5 seems to have broken (sorry!) the tests so I also fixed it in this PR by excluding jackson-databind from transitive dependencies.
I'm not so familiar with the Embulk API yet, so let me know if there's something wrong :)

joker1007 commented:
I'm sorry for the very late response.
I was too busy in the last 2 weeks. And so I forgot to watch this.

I have watched your commits and It looks good to me.
Thank you so much!
